extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java

```diff
@@ -0,0 +1,192 @@
+
+            status = ExecutionStatus.RUNNING;
+
+            // Note that because synchronizations can register other synchronizations
+            // we cannot use enhanced for loops as that could cause a concurrency exception
+            for (int i = 0; i < synchs.size(); i++) {
+                Synchronization sync = synchs.get(i);
+
+                try {
+                    sync.beforeCompletion();
+                } catch (Exception e) {
+                    if (LOGGER.isDebugEnabled()) {
+                        LOGGER.debugf(
+                                "The synchronization %s associated with tx key %s failed during beforeCompletion: %s",
+                                sync, tsr.getTransactionKey(), e.getMessage());
+                    }
+
+                    if (deferredThrowable == null) {
+                        deferredThrowable = e;
+
+                    }
+                }
+            }
+
+            status = ExecutionStatus.FINISHED;
+
+            if (deferredThrowable != null) {
+                /*
+                 * Cause the transaction to rollback. The underlying transaction manager will catch the runtime
+                 * exception and rethrow it when it does the rollback
+                 */
+                throw new RuntimeException(deferredThrowable);
+            }
+        }
+
+        @Override
+        public void afterCompletion(int status) {
+            // The list should be iterated in reverse order
+            for (int i = synchs.size(); i-- > 0;) {
+                synchs.get(i).afterCompletion(status);
+            }
+        }
+
+        // does packageName belong to this group of synchronizations
+        public boolean shouldAdd(String packageName) {
+            return !packagePrefix.isEmpty() && packageName.startsWith(packagePrefix);
+        }
+    }
+
+    private final List<SynchronizationGroup> synchGroups = new ArrayList<>();
+    private SynchronizationGroup otherSynchs;
+    private final TransactionSynchronizationRegistry tsr;
+
+    public AgroalOrderedLastSynchronizationList(
+            TransactionSynchronizationRegistryWrapper transactionSynchronizationRegistryWrapper) {
+
+        this.tsr = transactionSynchronizationRegistryWrapper;
+
+        for (var packagePrefix : PKG_PREFIXES) {
+            var synchronizationGroup = new SynchronizationGroup(packagePrefix);
+
+            synchGroups.add(synchronizationGroup);
+
+            if (packagePrefix.isEmpty()) {
+                otherSynchs = synchronizationGroup; // the catch-all group
+            }
+        }
+    }
+
+    /**
+     * Register an interposed synchronization. Note that synchronizations are not allowed if:
+     * <p>
+     *
+     * @param synchronization The synchronization to register
+     * @throws IllegalStateException if the transaction is in the wrong state:
+     *         <ol>
+     *         <li>the transaction has already prepared;
+     *         <li>the transaction is marked rollback only
```

Comment:
Are we throwing IllegalStateException because we want to fail-fast the marked for rollback only transaction?  If yes, the comment and the code below is fine.  If that isn't the intention then why are we throwing a failure when the transaction cannot succeed so not much else to do?

Reply:
It's a good question. I am throwing the exception here based on what we do in Narayana, see [1] and [2], where we throw an exception if it's marked for rollback. So, although I didn't write that Arjuna code, I would say it was done in order to fail fast.

[1] https://github.com/jbosstm/narayana/blob/7.0.0.Final/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java#L371
[2] https://github.com/jbosstm/narayana/blob/7.0.0.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TwoPhaseCoordinator.java#L116